examples/ct/structures.go

```diff
@@ -60,21 +60,60 @@ func signSCT(km crypto.KeyManager, t time.Time, sctData []byte) (ct.SignedCertif
 		Signature:  digitallySigned}, nil
 }
 
-// CreateV1SCTForCertificate builds and signs a V1 CT SCT for a certificate using the key held
-// by a key manager.
-func SignV1SCTForCertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.SignedCertificateTimestamp, error) {
-	return signSCT(km, t, cert.Raw)
+// CreateV1SCTForCertificate creates a MerkleTreeLeaf and builds and signs a V1 CT SCT for a certificate
+// using the key held by a key manager.
+func SignV1SCTForCertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.MerkleTreeLeaf, ct.SignedCertificateTimestamp, error) {
+	// Temp SCT for input to the serializer
+	sctInput := ct.SignedCertificateTimestamp{
+		SCTVersion: ct.V1,
+		Timestamp:  uint64(t.UnixNano() / millisPerNano), // spec uses millisecond timestamps
+		Extensions: ct.CTExtensions{}}
+
+	// Build up a MerkleTreeLeaf for the cert
+	timestampedEntry := ct.TimestampedEntry{Timestamp:sctInput.Timestamp, EntryType:ct.X509LogEntryType, X509Entry:cert.Raw}
+	leaf := ct.MerkleTreeLeaf{Version: ct.V1, LeafType:ct.TimestampedEntryLeafType, TimestampedEntry:timestampedEntry}
+
+	// Serialize SCT signature input to get the bytes that need to be signed
+	res, err := ct.SerializeSCTSignatureInput(sctInput, ct.LogEntry{Leaf:leaf})
+
+	if err != nil {
+		return ct.MerkleTreeLeaf{}, ct.SignedCertificateTimestamp{}, err
+	}
+
+	// Create a complete SCT including signature
+	sct, err := signSCT(km, t, res)
+
+	return leaf, sct, err
 }
 
 // CreateV1SCTForPrecertificate builds and signs a V1 CT SCT for a pre-certificate using the key
 // held by a key manager.
-func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.SignedCertificateTimestamp, error) {
+func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.MerkleTreeLeaf, ct.SignedCertificateTimestamp, error) {
```

Comment:
there's a lot of duplicated code.

can you extract a helper function that does the bulk of it, and to which these exported functions pass a func object for the bits in the middle that differ?